test/Driver/warnings-control.swift

```diff
@@ -9,7 +9,7 @@ func foo() -> Int {
 	let x = 1
 	var y = 2
 // DEFAULT:    warning: variable 'y' was never mutated; consider changing to 'let' constant
-// WERR:       error: variable 'y' was never mutated; consider changing to 'let' constant
+// WERR:       error: variable 'y' was never mutated; consider changing to 'let' constant [-warnings-as-errors]
```

Comment:
I would make the additional text a little more explicit, like `[upgraded from warning by -warnings-as-errors]`

Reply:
I used the Clang style (terse) because (1) it is already clear (in my opinion) and (2) it leaves room to add more stuff (say `[-warn-implicit-overrides, -warnings-as-errors]`) in a similar fashion without shuffling things around or making things too verbose. Thoughts?

Reply:
I think it might be a little too terse. Is `-warnings-as-errors` how you turn this off? Is it the reason it was emitted? Is it the reason it's an error instead of a warning? If we say something like `[-warn-implicit-overrides, -warnings-as-errors]`, do both of those flags have the same relationship to the diagnostic? And are these compiler flags at all, as opposed to something you might write in the source code?

Obviously if you know that `-warnings-as-errors` is a compiler flag and know or can guess what it does, you can probably figure out why it's there, but I don't think we should assume that users will have that background—the Xcode build setting is called "Treat Warnings as Errors" and searching the build settings for "warnings-as-errors" doesn't bring it up. Imagine a newbie developer who downloads an existing library, tries to build it on a newer Xcode, and sees `[-warnings-as-errors]` at the end of a diagnostic. Are they going to understand what it means?

It costs only a few words to make the message *much* more clear to someone who doesn't already know what you're talking about. What are we losing by adding it?

(P.S. In a world with perfect integration between all of our tools, I think what we would do is add a note with a fix-it to disable `-warnings-as-errors` in your build settings. Obviously that's not realistic, especially in the general case of an arbitrary build system.)

Reply:
Overall, I agree with the sentiment of helping a beginner developer out. If adding a few words can do that, sure I'm happy to add them. In this case, I don't feel too strongly one way or another; your point is convincing and I agree it would be good to add the few more words.

That does bring me to a broader question though: who are the diagnostics for? I feel like many of our diagnostics are very much _not_ beginner-friendly by this standard. For example, we use jargon like "non-nominal type" which a new developer may not know. The fact that we follow the Clang style for diagnostics (as @jrose-apple pointed out to me earlier), somewhat exacerbates this issue: we need to break the diagnostic up into a short description + additional notes, which means that we only have so much space before the short description gets "too long".

I'm not trying to engage in whatabout-ism here -- this is a genuine question. How heavily should be weighing such concerns going forward? Should we be slowly but steadily amending our existing diagnostics so that they hold up to this standard?

Reply:
This is essentially the problem I'm trying to solve with the "educational notes" I proposed as part of the extended diagnostics feature. That would let IDEs show a disclosure arrow or similar with a paragraph explaining what -warnings-as-errors is, etc instead of trying to fit it into the message line. 

Of course that doesn't help much right now because I haven't implemented it yet :). Until we've had a chance to improve diagnostic tooling, I like @brentdax's suggestion even though it doesn't exactly match the guidelines.